### PR TITLE
Use TypeScript (Requires #153)

### PR DESCRIPTION
**Pull request by Lucki**
Moved everything to TypeScript which will generate to JavaScript. This allows proper tooling like syntax checks, code highlights and docs additionally to type safety. It actually found quite a lot of smaller issues in the current code which are now fixed.
Because of this it's now required to have npm installed for building. I added a GitHub actions workflow which adds a build artifact to each push so easy access to nightly builds without manual building is available: [example](https://github.com/Lucki/RandomWallpaperGnome3/actions/workflows/build.yml)
One smaller caveat for TypeScript *currently* is that it can't properly type classes based on `GObject.registerClass()`. From what I've read it's currently an ongoing process to migrate this to proper ES6 classes in the GJS team.

Additionally, adapter can now request multiple new images at once and a lot of smaller fixes.

---

This already resolves comments from #141 :
* Libadwaita is now downloaded on demand
* I tried to preserve the JS history by moving the files.

~~Fun fact: While moving these commits from the other PR the majority of d6b5e6a was lost. I don't know why, I don't know how, but I had to rewrite almost everything.~~

Found the lost commit and inserted at the right place - I hope no other things were lost in the process…

### CODE REVIEW

**Comment by ArjixWasTaken** [comment]
Nice PR, don't mind if I fork it :^)

**Comment by ifl0w** [comment]
@Lucki finally, at the last stage :D ... Please rebase; I'll try to find the time to review during the next week!

**Comment by Lucki** [comment]
The first commits are still messy (and huge!). Don't hesitate to take your time :)

Some general things:
* Is the formatting for TS as you'd like? With eslint we can enforce and autoformat quite extensively.
    For JS, I'd like to stick the official GJS guidelines but still have to skimm through the generated files and create rules to add empty lines to improve readability.
* There are multiple timer fix mentions which I had some trouble with. I think I have it sorted out in the end but if the behavior doesn't meet your expectations I'll take another look.
* Is the behavior for the history with multiple displays as you'd like?
    I ended up with displaying the currently hovered preview on all displays and when setting a history element as active it will be assigned to the first display and others will be filled with random different history entries.

**Comment by ifl0w** [comment]
> The first commits are still messy (and huge!). Don't hesitate to take your time :)

Yeah, I noticed that! Sorry for taking so long; I didn't get very far till now. I hope to find the time over the next 2 weeks.

>  Is the formatting for TS as you'd like?

We can look at that later. Of course, I have preferences, but they never changed the output of a program ;).

> There are multiple timer fix mentions which I had some trouble with.

No worries, we'll find remaining bugs over time if there are any. I might have run into one, but I couldn't reproduce it. :shrug:  

> Is the behavior for the history with multiple displays as you'd like?

To be honest, we'll need to wait for the user feedback for this one. The best solution might be to let the user select the display for which to set the wallpaper. But that will require more UI and can be done at a later point.

**Comment by ifl0w** [comment]
Also, I noticed that Reddit now blocks requests without setting User-Agent headers. I'll collect any additions I do based on this PR on a [eparate branch](https://github.com/ifl0w/RandomWallpaperGnome3/tree/pr154_additions). 

Will open a PR once this PR here is merged :).

File: src/wallpaperController.ts
Comment (ifl0w) [inline src/wallpaperController.ts:382]: Can you briefly explain why you have to set the wallpaper in the new wallpaper manager instance and here again? Maybe we can reduce the places where we actually set picture URIs?

**Review by ifl0w — CHANGES_REQUESTED** [review]
Wow, once again, thanks for the hard work. This is a big change that touches on and resolves so many issues! Thanks!
I think your work is in pretty good shape already! Let's address/discuss the remaining comments to move these changes forward.

I'm testing the extension a bit to see if we need to improve something. Until now, I only noticed a few smaller things I would like to change in the UI, and I think I might have found a bug regarding hover previews while requesting new images, but I need to look into that further.

File: src/adapter/genericJson.ts
Comment (ifl0w) [reply src/adapter/genericJson.ts:48]: Of course, the same is true for the other sources.

File: src/adapter/baseAdapter.ts
Comment (Lucki) [reply src/adapter/baseAdapter.ts:10]: With `await Promise.all()` all requests will fire simultaneously for every display and multiple times in a row. First for requesting random images and then fetching the images. If the user doesn't like an image and presses the fetch button again the whole process repeats.
If all requests randomly decide to fetch from the same source I felt it couldn't hurt to try to space things out a bit.

File: src/wallpaperController.ts
Comment (Lucki) [reply src/wallpaperController.ts]: Yes, that's hardcoded in the merged image name

File: .eslintrc.yml
Comment (Lucki) [reply .eslintrc.yml:1]: This is the default config file and is used to format everything that is Typescript. It's not explicitly mentioned because it's loaded by default.

File: build.sh
Comment (Lucki) [reply build.sh:104]: Those are at the end of the rules section in `.eslintrc-gjs.yml`.
They are only `padding-line-between-statements:` to space things out in the generated JS files to adhere to the extension review guidelines.

https://gjs.guide/extensions/review-guidelines/review-guidelines.html#code-must-not-be-obfuscated

File: src/wallpaperController.ts
Comment (Lucki) [reply src/wallpaperController.ts:469]: > and from each, we would request 3 images

No, `_requestRandomAdapter()` returns for every adapter an `imageCount` to request.
For 3 monitors with 3 adapters the `imageCount` would be 1 for all three.
With 3 monitors and only 2 adapters the `imageCount` for 1 adapter would be 2 and for the other adapter 1.

``` ts
/**
 *
 * @param {RandomAdapterResult[]} array Array of already chosen adapter
 * @param {number} type Type of the source
 */
function _arrayIncludes(array: RandomAdapterResult[], type: number) {
    for (const element of array) {
        if (element.type === type)
            return element;
    }
    return null;
}

[…]

const availableAdapter = _arrayIncludes(randomAdapterResult, sourceType);
if (availableAdapter) {
    availableAdapter.imageCount++;
    continue;
}

[…]

randomAdapterResult.push({
    adapter: imageSourceAdapter,
    id: sourceID,
    type: sourceType,
    imageCount: 1,
});
```

File: src/history.ts
Comment (Lucki) [reply src/history.ts:31]: Well, there isn't alot carried over anyway.
The base ID is tied to extension name - I wouldn't change that.
I think the only problematic key that is re-used is `history`. We could only rename that key.

File: build.sh
Comment (Lucki) [reply build.sh:158]: Basically: yes.
But I figured it takes an unreasonable long time to do it's $stuff:
* The gh workflow runs steps individually and would re-run that setup step multiple times per build.
* `./build.sh build` speeds up testing alot.

File: types/ui/popupMenu.d.ts
Comment (Lucki) [reply types/ui/popupMenu.d.ts:1]: Well, that's the problem of Gnome having no API for extensions :laughing: 

Removing those definitions will effectively disable typing for most parts in `historyMenuElements.ts` and `randomWallpaperMenu.ts`.

Comment (ifl0w) [reply types/ui/popupMenu.d.ts:1]: Yeah, we won't remove them now anyways. I just hope that someone will eventually package and maintain the definitions as some npm package. ^^

File: src/adapter/baseAdapter.ts
Comment (ifl0w) [reply src/adapter/baseAdapter.ts:10]: Alright, thanks! Won't hurt, won't matter. :D Let's keep it this way.

File: src/wallpaperController.ts
Comment (ifl0w) [reply src/wallpaperController.ts:469]: Ah, right my bad! :+1:

File: .eslintrc.yml
Comment (ifl0w) [reply .eslintrc.yml:1]: Ah, that makes total sense! I didn't realize that `.eslintrc-gjs.yml` is applied to the transpiled JS.

File: build.sh
Comment (ifl0w) [reply build.sh:104]: Ah, thanks for the explanation. Can you add the reference to the guideline as a comment in the `.eslintrc-gjs.yml`, that would be great!

**Comment by Lucki** [comment]
My arch updated to Gnome 44 and this extension is now crashing somewhere in libmozjs before any debug output. Have to solve this before I can test things again.

File: src/wallpaperController.ts
Comment (Lucki) [reply src/wallpaperController.ts:393]: Added

File: build.sh
Comment (Lucki) [reply build.sh:104]: Added

File: types/ui/popupMenu.d.ts
Comment (Lucki) [reply types/ui/popupMenu.d.ts:1]: :pray:

File: src/settings.ts
Comment (Lucki) [reply src/settings.ts:154]: Do we? This is a copy of the internal schema finding mechanism, so assuming by the comments I expect system-wide installations having their schema located at a different path.

File: src/adapter/localFolder.ts
Comment (Lucki) [reply src/adapter/localFolder.ts:51]: > Should check for < count instead.

Currently, it only fails if no new images will be found (=== 0) this will escalate back upwards and results in no new images being set. If at least one new image is found it won't fail and fill missing images randomly from the history later on.
I can see that this will still fail when only one of multiple adapters fail. I should try to count the failing promises here.

But from your comment I assume you want to fail already if not all requested images can be fulfilled?

> Maybe also log a warning and return some partial results (this is just an idea).

Added log warning.
Partial results are already happening, see above.

File: src/history.ts
Comment (Lucki) [reply src/history.ts:81]: Because the compiler already knows about the exact type from the return statement.
Should I add a rule to force these?

**Comment by ifl0w** [comment]
> My arch updated to Gnome 44 and this extension is now crashing somewhere in libmozjs before any debug output. Have to solve this before I can test things again.

Dang, that's unfortunate. I might update later this week; maybe I can help find the issue if you haven't solved it until then. Let me know when you have any updates on it.

File: src/adapter/localFolder.ts
Comment (ifl0w) [reply src/adapter/localFolder.ts:51]: > But from your comment I assume you want to fail already if not all requested images can be fulfilled?
Basically yes, I would have expected it to fail just based on the function signature "request N images".

I think I'll have to take some more time to get more familiar with all changes and experiment with the code a bit.

File: src/settings.ts
Comment (ifl0w) [reply src/settings.ts:154]: We package the schemas folder with the app. I thought that any standard installation would have the packaged schema and we can expect this to folder to exist. 

But we can leave the code as is to be safe. Also, I don't know how this will be handled in the future when we eventually do not ship compiled schemas anymore.

File: src/adapter/baseAdapter.ts
Comment (ifl0w) [reply src/adapter/baseAdapter.ts:70]: Yeah, nevermind my bad. :sweat_smile:

File: src/history.ts
Comment (ifl0w) [reply src/history.ts:81]: I think it is generally a good thing to force return types for at least public methods so that any change to the return type creates "local" errors only (should result in easier-to-understand compile errors). 

Since this project didn't have any typing before, you can leave it as it is for now. :)

File: src/adapter/localFolder.ts
Comment (Lucki) [reply src/adapter/localFolder.ts:51]: Sure, I'll change it to this behavior. :+1:

File: src/settings.ts
Comment (Lucki) [reply src/settings.ts:154]: I had a look at the AUR package for gsconnect:
* The schema will be in `/usr/share/glib-2.0/schemas/….gschema.xml`
* The extension itself will be in `/usr/share/gnome-shell/extensions/…`

**Comment by Lucki** [comment]
Not solved yet. With `log('asd');` I tested a bit around and found that I can trigger the segfault even before the module loading promise by inserting a timeout to wait for. There's something strange going on in gjs + x11.

File: src/adapter/localFolder.ts
Comment (Lucki) [reply src/adapter/localFolder.ts]: [`44a6a76` (#154)](https://github.com/ifl0w/RandomWallpaperGnome3/pull/154/commits/44a6a76364b422e9b39d2119abe8dc964ddff73f)

File: src/history.ts
Comment (Lucki) [reply src/history.ts]: Forcing them now [`5718f98` (#154)](https://github.com/ifl0w/RandomWallpaperGnome3/pull/154/commits/5718f982e7fbe3b090cf36f2cfaaa7a5ec782d2d)

File: types/gtk4/adw/adw.d.ts
Comment (Lucki) [reply types/gtk4/adw/adw.d.ts]: Done [`aa7c1fb` (#154)](https://github.com/ifl0w/RandomWallpaperGnome3/pull/154/commits/aa7c1fb2515ac7fa75c4f25eb276a5be842dd060)

**Comment by Lucki** [comment]
https://gitlab.gnome.org/GNOME/gnome-shell/-/issues/6691

File: src/adapter/urlSource.ts
Comment (Lucki) [reply src/adapter/urlSource.ts]: Added a toggle, so the user can define when the URL is returning different images when requesting in a short amount time.

File: src/wallpaperController.ts
Comment (Lucki) [reply src/wallpaperController.ts:382]: This is the original place from your old code. Initially I gave HydraPaper some special treatment inline but while adding SuperPaper I moved those to the according manager.
They behave slightly different in what settings they set on their own and how we have to handle these.

This code here has some special treatment as well for hover previews. When returning to the current wallpaper it could be an already merged wallpaper which then needs a different zoom option. That's the search in the name for 'merged_wallpaper' and 'cli-[a|b]'.

For reducing complexity I see the following options:
* We could introduce a new default wallpaperManager instance which would handle these default cases. Setting the URIs would be in similar places then.
* We could ditch the lockscreen setting. That would remove the need for special treatments. I honestly never saw that showing up anywhere. But the gnome settings app also sets the lockscreen value on wallpaper change so it should show up somewhere?

File: build.sh
Comment (Lucki) [reply build.sh]: I was able to remove all of them except the TS bug. :+1:

File: .github/workflows/build.yml
Comment (Lucki) [reply .github/workflows/build.yml]: Manually downloading the deb from Kinetic (22.10) seems to work. [`cd1e195` (#154)](https://github.com/ifl0w/RandomWallpaperGnome3/pull/154/commits/cd1e195024ee66d91bdf454a4b8f70ea591a2bf4)

File: src/adapter/localFolder.ts
Comment (Lucki) [reply src/adapter/localFolder.ts]: I'm not entirely sure if this is what you had in mind but it's definitely nicer this way: [`b174bb5` (#154)](https://github.com/ifl0w/RandomWallpaperGnome3/pull/154/commits/b174bb5c2fb6649f73e9a3a1f23baea058897c2f)

File: src/settings.ts
Comment (ifl0w) [reply src/settings.ts:296]: Ah, sorry, now I understand your point!

File: .github/workflows/build.yml
Comment (ifl0w) [reply .github/workflows/build.yml]: Significantly nicer solution! Thanks!

File: src/adapter/localFolder.ts
Comment (ifl0w) [reply src/adapter/localFolder.ts]: That was exactly what I had in mind! Thanks!

**Comment by ifl0w** [comment]
Since the fix for the startup issues is already merged in gnome, can we remove the workaround here?

**Comment by Lucki** [comment]
I would at least wait for the fixing commit to show up in *any* release before removing it.

File: src/wallpaperController.ts
Comment (Lucki) [reply src/wallpaperController.ts:420]: I added the first option locally. Have to give it some tests before I'll push here tough.

**Comment by Lucki** [comment]
The fix appeared in https://gitlab.gnome.org/GNOME/gnome-shell/-/tags/45.alpha (`!2781`) so removing the workaround will bump the minimum required version to Gnome 45 if it won't get backported later on.

File: src/wallpaperController.ts
Comment (ifl0w) [reply src/wallpaperController.ts]: > I honestly never saw that showing up anywhere. But the gnome settings app also sets the lockscreen value on wallpaper change so it should show up somewhere?

I don't know if it is actually used anywhere in gnome anymore. It used to show up when locking/unlocking your screen iirc. Now there is always the blurred background I think... :shrug:

**Review by ifl0w — APPROVED** [review]
Looks good to me! 

Please fix the one commit message, and then let's merge it.

File: src/utils.ts
Comment (Lucki) [reply src/utils.ts]: I even called it fixup but already pushed and did not want to force push while you're possibly in a review state.

File: package.json
Comment (Lucki) [reply package.json]: Pinned them to the current version

**Pull request merged**
